Fix usrsctp usage in Rust

ibc opened the pull request:
**WIP**

Fixes https://github.com/versatica/mediasoup/issues/1352

### Details

- Basically as described in the ticket. But not everything is done at all.
- Also, I'm testing this in Node by using UV async stuff (which doesn't make sense in mediasoup for Node but anyway).

### TODO

- None of these changes should take effect when in Node, so we need to pass (or to NOT pass) some `define` only from Rust to enable this in the C++ code. We don't want to deal with UV async stuff when in Node because it's not needed at all, so let's see how to do it.

- Missing thread X to initialize usrsctp and run the `Checker` singleton. And many other things.

ibc committed:
Fix usrsctp usage in Rust

**WIP**

Fixes https://github.com/versatica/mediasoup/issues/1352

### Details

- Basically as described in the ticket. But not everything is done at all.
- Also, I'm testing this in Node by using UV async stuff (which doesn't make sense in mediasoup for Node but anyway).

### TODO

- None of these changes should take effect when in Node, so we need to pass (or to NOT pass) some `define` only from Rust to enable this in the C++ code. We don't want to deal with UV async stuff when in Node because it's not needed at all, so let's see how to do it.

- Missing thread X to initialize usrsctp and run the `Checker` singleton. And many other things.

- Crash when a `SctpAssociation` is closed. I think it's because somehow the `onAsync` callback is invoked asynchronously (of course) so when it calls `sctpAssociation->OnUsrSctpSendSctpData()` it happens that such a `SctpAssociation` has already been freed. Not sure how to resolve it. Here the logs:
  ```
  mediasoup:Transport close() +18s
  mediasoup:Channel request() [method:ROUTER_CLOSE_TRANSPORT] +8s
  mediasoup:Producer transportClosed() +19s
  mediasoup:DataProducer transportClosed() +18s
  mediasoup:DataProducer transportClosed() +0ms
  mediasoup:Transport close() +1ms
  mediasoup:Channel request() [method:ROUTER_CLOSE_TRANSPORT] +1ms
  mediasoup:Consumer transportClosed() +19s
  mediasoup:DataConsumer transportClosed() +18s
  mediasoup:DataConsumer transportClosed() +1ms
  mediasoup:Channel [pid:98040] RTC::SctpAssociation::ResetSctpStream() | SCTP_RESET_STREAMS sent [streamId:1] +1ms
  mediasoup:Channel request succeeded [method:ROUTER_CLOSE_TRANSPORT, id:39] +0ms
  DepUsrSCTP::onAsync() | ---------- onAsync!!
  DepUsrSCTP::onAsync() | ---------- onAsync, sending SCTP data!!
  mediasoup:Channel Producer Channel ended by the worker process +1ms
  mediasoup:ERROR:Worker worker process died unexpectedly [pid:98040, code:null, signal:SIGSEGV] +0ms
  ```

ibc committed:
debugging logs

ibc committed:
Delete mem copied data

ibc committed:
more logs

ibc committed:
debugging

ibc committed:
properly close the uv_async_t handle

ibc committed:
test-node-sctp.ts is failing, add debug

ibc commented:
### Issue 1: test-node-sctp.ts fails

I've added some console logs: https://github.com/versatica/mediasoup/pull/1353/commits/069f78e62c618afa37622e119b4ea0d522fc9da1

The test fails because first sent message is not later received by the data consumer:

```
npx jest --testPathPattern node/src/test/test-node-sctp.ts
  console.log
    TODO: Revert numMessages to 200

      at node/src/test/test-node-sctp.ts:116:10

  console.log
    ---- sending id 1

      at node/src/test/test-node-sctp.ts:143:12

  console.log
    ---- sending id 2

      at node/src/test/test-node-sctp.ts:143:12

  console.log
    ---- sending id 3

      at node/src/test/test-node-sctp.ts:143:12

  console.log
    ---- received id 2

      at SCTPStreamReadable.<anonymous> (node/src/test/test-node-sctp.ts:175:13)

 FAIL  test/test-node-sctp.ts (7.075 s)
  ✕ ordered DataProducer delivers all SCTP messages to the DataConsumer (4267 ms)

  ● ordered DataProducer delivers all SCTP messages to the DataConsumer

    id 2 in message should match numReceivedMessages 1

      177 | 				if (id !== numReceivedMessages) {
      178 | 					reject(
    > 179 | 						new Error(
          | 						^
      180 | 							`id ${id} in message should match numReceivedMessages ${numReceivedMessages}`
      181 | 						)
      182 | 					);

      at SCTPStreamReadable.<anonymous> (node/src/test/test-node-sctp.ts:179:7)
```


**UPDATE:** Issue found. Problem is that when usrsctp send callback is called, there we store sending data into a map and then invoke uv async (which will happen time later) and such a uv async will read from that storage to send the data. Problem is that, if two sequential messages must be sent to a peer, the second one will override the first one in the storage so when uv async callback is executed it will only read the second one. Here the issue (see how `onSendSctpData()` is called twice sequentially):

```
---- test | sending id 1
---- test | sending id 2
RTC::Transport::OnSctpAssociationMessageReceived() | ------ [streamId:123, ppid:53, msg:'1']
RTC::DataProducer::ReceiveMessage() | ------ [msg:'1']
RTC::Router::OnTransportDataProducerMessageReceived() | ------ [ppid:53, msg:'1']
RTC::DataConsumer::SendMessage() | ------ [ppid:53, msg:'1']
RTC::Transport::OnDataConsumerSendMessage() | ------ [ppid:53, msg:'1']
RTC::PlainTransport::SendMessage() | ------ [ppid:53, msg:'1']
RTC::SctpAssociation::SendSctpMessage() | ------ [ppid:53, msg:'1']
DepUsrSCTP::onSendSctpData() | ****
DepUsrSCTP::SendSctpData() | **** previous store.data: 0x0 [len:0]
DepUsrSCTP::onSendSctpData() | ****
DepUsrSCTP::SendSctpData() | **** previous store.data: 0x600001a88040 [len:32]
DepUsrSCTP::onAsync() | ****
RTC::Transport::OnSctpAssociationMessageReceived() | ------ [streamId:123, ppid:53, msg:'2']
RTC::DataProducer::ReceiveMessage() | ------ [msg:'2']
RTC::Router::OnTransportDataProducerMessageReceived() | ------ [ppid:53, msg:'2']
RTC::DataConsumer::SendMessage() | ------ [ppid:53, msg:'2']
RTC::Transport::OnDataConsumerSendMessage() | ------ [ppid:53, msg:'2']
RTC::PlainTransport::SendMessage() | ------ [ppid:53, msg:'2']
RTC::SctpAssociation::SendSctpMessage() | ------ [ppid:53, msg:'2']
DepUsrSCTP::onSendSctpData() | ****
DepUsrSCTP::SendSctpData() | **** previous store.data: 0x0 [len:0]
DepUsrSCTP::onAsync() | ****
---- test | received id 2

 RUNS  test/test-node-sctp.ts
 FAIL  test/test-node-sctp.ts
  ✕ ordered DataProducer delivers all SCTP messages to the DataConsumer (1645 ms)

  ● ordered DataProducer delivers all SCTP messages to the DataConsumer

    id 2 in message should match numReceivedMessages 1
```

Solution is: Store all pending messages to be sent in a container.

jmillan commented:
Maybe related to this?:

https://docs.libuv.org/en/v1.x/async.html

`libuv will coalesce calls to uv_async_send(), that is, not every call to it will yield an execution of the callback. For example: if uv_async_send() is called 5 times in a row before the callback is called, the callback will only be called once. If uv_async_send() is called again after the callback was called, it will be called again.`

ibc committed:
Use a vector to store sending messages and clean it once messages are sent in async cb

ibc commented:
Ok, so I've solved the problem as follows: https://github.com/versatica/mediasoup/pull/1353/commits/1a25bedd503f2a96f3d745f7f9af203e02688ab9

Test now works (for simplicity is temporary reduced to just 2 messages) and all logs are good:

```
DepUsrSCTP::SendSctpDataStore() | ---------- store constructor [fooId:1]
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:0]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:1, items.size:1]
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:1]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:1, data:0x6000014fc620]
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:2, items.size:2]
DepUsrSCTP::onAsync() | **** onAsync() called
DepUsrSCTP::onAsync() | ------------- sending pening messages [items.size:2]
DepUsrSCTP::ClearItems() | ---------- store ClearItems() [fooId:1, items.size():2]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:2, data:0x7f78ff705690]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:1, data:0x6000014fc620]
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:0]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:3, items.size:1]
DepUsrSCTP::onAsync() | **** onAsync() called
DepUsrSCTP::onAsync() | ------------- sending pening messages [items.size:1]
DepUsrSCTP::ClearItems() | ---------- store ClearItems() [fooId:1, items.size():1]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:3, data:0x600003df8090]
TODO: Revert numMessages to 200
---- test | sending id 1
---- test | sending id 2
RTC::Transport::OnSctpAssociationMessageReceived() | ------ [streamId:123, ppid:53, msg:'1']
RTC::DataProducer::ReceiveMessage() | ------ [msg:'1']
RTC::DataConsumer::SendMessage() | ------ [ppid:53, msg:'1']
RTC::SctpAssociation::SendSctpMessage() | ------ [ppid:53, msg:'1']
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:0]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:4, items.size:1]
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:1]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:5, items.size:2]
DepUsrSCTP::onAsync() | **** onAsync() called
DepUsrSCTP::onAsync() | ------------- sending pening messages [items.size:2]
DepUsrSCTP::ClearItems() | ---------- store ClearItems() [fooId:1, items.size():2]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:5, data:0x600003fe0c00]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:4, data:0x600003fe08e0]
---- test | received id 1
RTC::Transport::OnSctpAssociationMessageReceived() | ------ [streamId:123, ppid:53, msg:'2']
RTC::DataProducer::ReceiveMessage() | ------ [msg:'2']
RTC::DataConsumer::SendMessage() | ------ [ppid:53, msg:'2']
RTC::SctpAssociation::SendSctpMessage() | ------ [ppid:53, msg:'2']
DepUsrSCTP::onSendSctpData() | **** onSendSctpData() called
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back()... [items.size:0]
DepUsrSCTP::SendSctpDataItem() | ---------- item constructor
DepUsrSCTP::SendSctpData() | -------- store.items.emplace_back() DONE [item.fooId:6, items.size:1]
DepUsrSCTP::onAsync() | **** onAsync() called
DepUsrSCTP::onAsync() | ------------- sending pening messages [items.size:1]
DepUsrSCTP::ClearItems() | ---------- store ClearItems() [fooId:1, items.size():1]
DepUsrSCTP::~SendSctpDataItem() | ---------- item destructor [fooId:6, data:0x600003ffeec0]
---- test | received id 2

 PASS  test/test-node-sctp.ts (5.395 s)
  ✓ ordered DataProducer delivers all SCTP messages to the DataConsumer (1608 ms)
```

ibc committed:
remove debugging changes in test file

ibc committed:
cosmetic

ibc commented:
And **OF COURSE** a double free bug that just happens in Rust in CI in Ubuntu and not in MacOS, to make me spend yet another 2 extra days on this: https://github.com/versatica/mediasoup/actions/runs/8172985292/job/22344414084?pr=1353

jmillan commented:
Are you making sure that threadA is not writing to a `SendSctpDataStore` at the when at the same time the `uv` main thread is reading it?

ibc commented:
> Are you making sure that threadA is not writing to a `SendSctpDataStore` at the when at the same time the `uv` main thread is reading it?

ClassDestroy() method should also use the mutex since it clears the storage map so every store is deallocated so it deletes all items in the items array. It maybe that.

ibc commented:
Also take into account that this PR doesn't yet implement the thing about running the Checker in a separate thread.

ibc commented:
> > Are you making sure that threadA is not writing to a `SendSctpDataStore` at the when at the same time the `uv` main thread is reading it?
> 
> ClassDestroy() method should also use the mutex since it clears the storage map so every store is deallocated so it deletes all items in the items array. It maybe that.

No, false alarm, it already calls `const std::lock_guard<std::mutex> lock(GlobalSyncMutex);`.

ibc commented:
> And **OF COURSE** a double free bug that just happens in Rust in CI in Ubuntu and not in MacOS, to make me spend yet another 2 extra days on this: https://github.com/versatica/mediasoup/actions/runs/8172985292/job/22344414084?pr=1353

I got a different crash running `cargo test` in mac:

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000010
Exception Codes:       0x0000000000000001, 0x0000000000000010

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [23080]

VM Region Info: 0x10 is not in any region.  Bytes before following region: 140737487462384
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      shared memory            7ffffff26000-7ffffff27000 [    4K] r-x/r-x SM=SHM  

Thread 0 Crashed:: main Dispatch queue: com.apple.main-thread
0   mediasoup-66251c5bfb485c86    	       0x10c7ddddb Channel::ChannelSocket::SendLog(char const*, unsigned int) + 27
1   mediasoup-66251c5bfb485c86    	       0x10c79fdf0 DepUsrSCTP::SendSctpDataStore::~SendSctpDataStore() + 144
2   mediasoup-66251c5bfb485c86    	       0x10c79ff65 DepUsrSCTP::SendSctpDataStore::~SendSctpDataStore() + 21
3   mediasoup-66251c5bfb485c86    	       0x10c7a0a69 std::__1::pair<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>::~pair() + 25
4   mediasoup-66251c5bfb485c86    	       0x10c7a0a45 std::__1::pair<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>::~pair() + 21
5   mediasoup-66251c5bfb485c86    	       0x10c7a0a29 void std::__1::allocator_traits<std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::destroy[abi:v160006]<std::__1::pair<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>, void, void>(std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>&, std::__1::pair<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>*) + 25
6   mediasoup-66251c5bfb485c86    	       0x10c7a09fd void absl::lts_20230802::container_internal::map_slot_policy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>::destroy<std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>(std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>*, absl::lts_20230802::container_internal::map_slot_type<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>*) + 29
7   mediasoup-66251c5bfb485c86    	       0x10c7a09cd void absl::lts_20230802::container_internal::FlatHashMapPolicy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>::destroy<std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>(std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>*, absl::lts_20230802::container_internal::map_slot_type<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>*) + 29
8   mediasoup-66251c5bfb485c86    	       0x10c7a099d void absl::lts_20230802::container_internal::common_policy_traits<absl::lts_20230802::container_internal::FlatHashMapPolicy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>, void>::destroy<std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>(std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>*, absl::lts_20230802::container_internal::map_slot_type<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>*) + 29
9   mediasoup-66251c5bfb485c86    	       0x10c7a07a8 absl::lts_20230802::container_internal::raw_hash_set<absl::lts_20230802::container_internal::FlatHashMapPolicy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Hash, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Eq, std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::destroy_slots() + 136
10  mediasoup-66251c5bfb485c86    	       0x10c7a0633 absl::lts_20230802::container_internal::raw_hash_set<absl::lts_20230802::container_internal::FlatHashMapPolicy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Hash, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Eq, std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::~raw_hash_set() + 67
11  mediasoup-66251c5bfb485c86    	       0x10c7a05e5 absl::lts_20230802::container_internal::raw_hash_map<absl::lts_20230802::container_internal::FlatHashMapPolicy<uv_async_s const*, DepUsrSCTP::SendSctpDataStore>, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Hash, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Eq, std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::~raw_hash_map() + 21
12  mediasoup-66251c5bfb485c86    	       0x10c7a05c5 absl::lts_20230802::flat_hash_map<uv_async_s const*, DepUsrSCTP::SendSctpDataStore, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Hash, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Eq, std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::~flat_hash_map() + 21
13  mediasoup-66251c5bfb485c86    	       0x10c79d855 absl::lts_20230802::flat_hash_map<uv_async_s const*, DepUsrSCTP::SendSctpDataStore, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Hash, absl::lts_20230802::container_internal::HashEq<uv_async_s const*, void>::Eq, std::__1::allocator<std::__1::pair<uv_async_s const* const, DepUsrSCTP::SendSctpDataStore>>>::~flat_hash_map() + 21
14  libsystem_c.dylib             	    0x7ff80d6a4ba1 __cxa_finalize_ranges + 409
15  libsystem_c.dylib             	    0x7ff80d6a49bb exit + 35
16  libdyld.dylib                 	    0x7ff80d7ea8d3 dyld4::LibSystemHelpers::exit(int) const + 11
17  dyld                          	    0x7ff80d479458 start + 1960

Thread 1:
0   libsystem_kernel.dylib        	    0x7ff80d79451e semaphore_wait_trap + 10
1   libdispatch.dylib             	    0x7ff80d632502 _dispatch_sema4_wait + 16
2   libdispatch.dylib             	    0x7ff80d6329f5 _dispatch_semaphore_wait_slow + 98
3   mediasoup-66251c5bfb485c86    	       0x10ca3dca0 std::thread::park::h5e1b432e40d2a3d0 + 64
4   mediasoup-66251c5bfb485c86    	       0x10bedc793 std::sync::mpmc::context::Context::wait_until::hb9e79df896b2a063 + 403 (context.rs:139)
5   mediasoup-66251c5bfb485c86    	       0x10bedf1a2 std::sync::mpmc::list::Channel$LT$T$GT$::recv::_$u7b$$u7b$closure$u7d$$u7d$::h85c30bb20d4c8cd8 + 210 (list.rs:444)
6   mediasoup-66251c5bfb485c86    	       0x10bcddee4 std::sync::mpmc::context::Context::with::_$u7b$$u7b$closure$u7d$$u7d$::h8199c836e5f91c16 + 196 (context.rs:50)
7   mediasoup-66251c5bfb485c86    	       0x10bcddff8 std::sync::mpmc::context::Context::with::_$u7b$$u7b$closure$u7d$$u7d$::h833faaf98f9fbe3a + 264 (context.rs:58)
8   mediasoup-66251c5bfb485c86    	       0x10bd32e47 std::thread::local::LocalKey$LT$T$GT$::try_with::h3048f0533ea5520b + 247 (local.rs:270)
9   mediasoup-66251c5bfb485c86    	       0x10bcdd9af std::sync::mpmc::context::Context::with::h7a684c5b529ce549 + 63 (context.rs:53)
10  mediasoup-66251c5bfb485c86    	       0x10bedf091 std::sync::mpmc::list::Channel$LT$T$GT$::recv::hdcf783a177a61906 + 369 (list.rs:434)
11  mediasoup-66251c5bfb485c86    	       0x10bc374c3 std::sync::mpmc::Receiver$LT$T$GT$::recv::hda542cd8bafb090e + 131 (mod.rs:307)
12  mediasoup-66251c5bfb485c86    	       0x10bcb0361 std::sync::mpsc::Receiver$LT$T$GT$::recv::h5e2db09bc8ce0a62 + 17 (mod.rs:860)
13  mediasoup-66251c5bfb485c86    	       0x10bdbdf7f _$LT$mediasoup..worker_manager..Inner$u20$as$u20$core..ops..drop..Drop$GT$::drop::hee0be456b28462b5 + 447 (worker_manager.rs:39)
14  mediasoup-66251c5bfb485c86    	       0x10bd94a45 core::ptr::drop_in_place$LT$mediasoup..worker_manager..Inner$GT$::hb0e852432b141118 + 21 (mod.rs:497)
15  mediasoup-66251c5bfb485c86    	       0x10bca3d70 alloc::sync::Arc$LT$T$GT$::drop_slow::h8f411e3bb305f379 + 32 (sync.rs:1263)
16  mediasoup-66251c5bfb485c86    	       0x10bd9ffa7 _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h05d259f83087abfe + 103 (sync.rs:1897)
17  mediasoup-66251c5bfb485c86    	       0x10bd9ab51 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$mediasoup..worker_manager..Inner$GT$$GT$::h21006cf990561a66 + 17 (mod.rs:497)
18  mediasoup-66251c5bfb485c86    	       0x10bd96901 core::ptr::drop_in_place$LT$mediasoup..worker_manager..WorkerManager$GT$::hd1f65ec374c9433d + 17 (mod.rs:497)
19  mediasoup-66251c5bfb485c86    	       0x10bd93e35 core::ptr::drop_in_place$LT$mediasoup..worker..Inner$GT$::h5a8384bf3d72b9fb + 277 (mod.rs:497)
20  mediasoup-66251c5bfb485c86    	       0x10bca3a90 alloc::sync::Arc$LT$T$GT$::drop_slow::h79087e6c5fe2c30c + 32 (sync.rs:1263)
21  mediasoup-66251c5bfb485c86    	       0x10bda2e27 _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hc11d9995d60ac333 + 103 (sync.rs:1897)
22  mediasoup-66251c5bfb485c86    	       0x10bd99261 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$mediasoup..worker..Inner$GT$$GT$::haad5c6b67b3454e0 + 17 (mod.rs:497)
23  mediasoup-66251c5bfb485c86    	       0x10bd93fc1 core::ptr::drop_in_place$LT$mediasoup..worker..Worker$GT$::h6a00775293fc8c0c + 17 (mod.rs:497)
24  mediasoup-66251c5bfb485c86    	       0x10bd93cd9 core::ptr::drop_in_place$LT$mediasoup..router..Inner$GT$::h942a7fdd3803ed06 + 409 (mod.rs:497)
25  mediasoup-66251c5bfb485c86    	       0x10bca35e0 alloc::sync::Arc$LT$T$GT$::drop_slow::h53b41c1cdab84c70 + 32 (sync.rs:1263)
26  mediasoup-66251c5bfb485c86    	       0x10bda0327 _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h1299bb24131f8738 + 103 (sync.rs:1897)
27  mediasoup-66251c5bfb485c86    	       0x10bd99241 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$mediasoup..router..Inner$GT$$GT$::h52d49354ff02fe37 + 17 (mod.rs:497)
28  mediasoup-66251c5bfb485c86    	       0x10bd93fa1 core::ptr::drop_in_place$LT$mediasoup..router..Router$GT$::h1eb51bf5d762bfa4 + 17 (mod.rs:497)
29  mediasoup-66251c5bfb485c86    	       0x10bd9758f core::ptr::drop_in_place$LT$mediasoup..router..webrtc_transport..Inner$GT$::h43c71abd6b38f469 + 415 (mod.rs:497)
30  mediasoup-66251c5bfb485c86    	       0x10bca3870 alloc::sync::Arc$LT$T$GT$::drop_slow::h68f205e85b17ef92 + 32 (sync.rs:1263)
31  mediasoup-66251c5bfb485c86    	       0x10bda3c27 _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hf67587d2568d0006 + 103 (sync.rs:1897)
32  mediasoup-66251c5bfb485c86    	       0x10bd9bfd1 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$mediasoup..router..webrtc_transport..Inner$GT$$GT$::hb0b82dfc86ac1f95 + 17 (mod.rs:497)
33  mediasoup-66251c5bfb485c86    	       0x10bd9a381 core::ptr::drop_in_place$LT$mediasoup..router..webrtc_transport..WebRtcTransport$GT$::h77030e6465e3821e + 17 (mod.rs:497)
34  mediasoup-66251c5bfb485c86    	       0x10bd98f52 core::ptr::drop_in_place$LT$dyn$u20$mediasoup..router..transport..Transport$GT$::h09c226a504180469 + 18 (mod.rs:497)
35  mediasoup-66251c5bfb485c86    	       0x10bca449a alloc::sync::Arc$LT$T$GT$::drop_slow::hbdaa5299bc2765ea + 58 (sync.rs:1263)
36  mediasoup-66251c5bfb485c86    	       0x10bda2f2f _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hc4f1c69d8c18abdf + 111 (sync.rs:1897)
37  mediasoup-66251c5bfb485c86    	       0x10bd9d411 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$dyn$u20$mediasoup..router..transport..Transport$GT$$GT$::hbed9f79d0d68d61f + 17 (mod.rs:497)
38  mediasoup-66251c5bfb485c86    	       0x10bd9513b core::ptr::drop_in_place$LT$mediasoup..router..consumer..Inner$GT$::h0771200344ca7506 + 491 (mod.rs:497)
39  mediasoup-66251c5bfb485c86    	       0x10bca38b0 alloc::sync::Arc$LT$T$GT$::drop_slow::h6a1c7123abf76371 + 32 (sync.rs:1263)
40  mediasoup-66251c5bfb485c86    	       0x10bda1727 _$LT$alloc..sync..Arc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h6a75c7e368fd7050 + 103 (sync.rs:1897)
41  mediasoup-66251c5bfb485c86    	       0x10bd9aef1 core::ptr::drop_in_place$LT$alloc..sync..Arc$LT$mediasoup..router..consumer..Inner$GT$$GT$::h6565bb57939aec22 + 17 (mod.rs:497)
42  mediasoup-66251c5bfb485c86    	       0x10be63f40 mediasoup::router::consumer::Consumer::new::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h526886ea594fc192 + 128 (consumer.rs:817)
43  mediasoup-66251c5bfb485c86    	       0x10bd08348 async_executor::Executor::spawn::_$u7b$$u7b$closure$u7d$$u7d$::h825ecaeb4660303d + 376 (lib.rs:145)
44  mediasoup-66251c5bfb485c86    	       0x10bde06e6 async_task::raw::RawTask$LT$F$C$T$C$S$C$M$GT$::run::h20ff3f18aad9af6d + 966 (raw.rs:563)
45  mediasoup-66251c5bfb485c86    	       0x10c207345 async_task::runnable::Runnable$LT$M$GT$::run::hfab01958ac2123c0 + 117
46  mediasoup-66251c5bfb485c86    	       0x10bd02046 async_executor::Executor::run::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h0229ff2a89242fef + 790 (lib.rs:236)
47  mediasoup-66251c5bfb485c86    	       0x10bcbaae8 _$LT$futures_lite..future..Or$LT$F1$C$F2$GT$$u20$as$u20$core..future..future..Future$GT$::poll::hd96f6a372ace1ca4 + 104 (future.rs:529)
48  mediasoup-66251c5bfb485c86    	       0x10bd01c0a async_executor::Executor::run::_$u7b$$u7b$closure$u7d$$u7d$::h5fa344a4260228ae + 586 (lib.rs:243)
49  mediasoup-66251c5bfb485c86    	       0x10bcb9e84 futures_lite::future::block_on::_$u7b$$u7b$closure$u7d$$u7d$::hc234b6f9b0c6e07f + 372 (future.rs:89)
50  mediasoup-66251c5bfb485c86    	       0x10bd34517 std::thread::local::LocalKey$LT$T$GT$::try_with::hd8d6482ef34b0ae3 + 247 (local.rs:270)
51  mediasoup-66251c5bfb485c86    	       0x10bd32525 std::thread::local::LocalKey$LT$T$GT$::with::h777e1829f914babc + 21 (local.rs:246)
52  mediasoup-66251c5bfb485c86    	       0x10bcb71f1 futures_lite::future::block_on::hd041d5b54e6a1aa7 + 129 (future.rs:79)
53  mediasoup-66251c5bfb485c86    	       0x10bbad3e8 mediasoup::worker_manager::WorkerManager::new::_$u7b$$u7b$closure$u7d$$u7d$::h16e4d4028c474547 + 120 (worker_manager.rs:94)
54  mediasoup-66251c5bfb485c86    	       0x10bd209ed std::sys_common::backtrace::__rust_begin_short_backtrace::hf31f30424128813c + 13 (backtrace.rs:135)
55  mediasoup-66251c5bfb485c86    	       0x10be62898 std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h1dd7bd15e29e8510 + 40 (mod.rs:529)
56  mediasoup-66251c5bfb485c86    	       0x10bc81a18 _$LT$core..panic..unwind_safe..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h233e942267404bf1 + 40 (unwind_safe.rs:271)
57  mediasoup-66251c5bfb485c86    	       0x10bd1e9dc std::panicking::try::do_call::h3d77e921ef020bee + 60 (panicking.rs:500)
58  mediasoup-66251c5bfb485c86    	       0x10bd2092d __rust_try + 29
59  mediasoup-66251c5bfb485c86    	       0x10bd1d8b6 std::panicking::try::h8ed850b51d367ea1 + 86 (panicking.rs:464)
60  mediasoup-66251c5bfb485c86    	       0x10be626e8 std::panic::catch_unwind::he040877442ec4da7 + 5 (panic.rs:142) [inlined]
61  mediasoup-66251c5bfb485c86    	       0x10be626e8 std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::h735b7425c3f644bc + 504 (mod.rs:528)
62  mediasoup-66251c5bfb485c86    	       0x10bd72a81 core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h081ff51a65d08564 + 17 (function.rs:250)
63  mediasoup-66251c5bfb485c86    	       0x10ca4b2e9 std::sys::unix::thread::Thread::new::thread_start::ha4a6ad20828e5661 + 41
64  libsystem_pthread.dylib       	    0x7ff80d7d31d3 _pthread_start + 125
65  libsystem_pthread.dylib       	    0x7ff80d7cebd3 thread_start + 15

Thread 2:: mediasoup-worker-6025bfba-49de-419c-bf16-a1f10a41bd2b
0   libsystem_kernel.dylib        	    0x7ff80d7991de kevent + 10
1   mediasoup-66251c5bfb485c86    	       0x10c6ebca2 uv__io_poll + 2002
2   mediasoup-66251c5bfb485c86    	       0x10c6dfc06 uv_run + 326
3   mediasoup-66251c5bfb485c86    	       0x10c79cff5 DepLibUV::RunLoop() + 117
4   mediasoup-66251c5bfb485c86    	       0x10c7b5c4a Worker::Worker(Channel::ChannelSocket*) + 458
5   mediasoup-66251c5bfb485c86    	       0x10c7b5fbd Worker::Worker(Channel::ChannelSocket*) + 29
6   mediasoup-66251c5bfb485c86    	       0x10c7990fb mediasoup_worker_run + 1675
7   mediasoup-66251c5bfb485c86    	       0x10bea7f3f mediasoup::worker::utils::run_worker_with_channels::_$u7b$$u7b$closure$u7d$$u7d$::h09ce28fe494011ac + 1231 (utils.rs:77)
8   mediasoup-66251c5bfb485c86    	       0x10bd209ad std::sys_common::backtrace::__rust_begin_short_backtrace::h1dea32572cf1056e + 13 (backtrace.rs:135)
9   mediasoup-66251c5bfb485c86    	       0x10be628cb std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h97d78e80c9c12ecb + 43 (mod.rs:529)
10  mediasoup-66251c5bfb485c86    	       0x10bc81c4b _$LT$core..panic..unwind_safe..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h9ae2cecdcdbc05df + 43 (unwind_safe.rs:271)
11  mediasoup-66251c5bfb485c86    	       0x10bd1eaef std::panicking::try::do_call::h5ac74dcda4157391 + 63 (panicking.rs:500)
12  mediasoup-66251c5bfb485c86    	       0x10bd2092d __rust_try + 29
13  mediasoup-66251c5bfb485c86    	       0x10bd1d94c std::panicking::try::h95e60b68bdcb7520 + 92 (panicking.rs:464)
14  mediasoup-66251c5bfb485c86    	       0x10be62367 std::panic::catch_unwind::h4c18e6cce43153fe + 5 (panic.rs:142) [inlined]
15  mediasoup-66251c5bfb485c86    	       0x10be62367 std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::h3d9bc5a6e16d5de0 + 503 (mod.rs:528)
16  mediasoup-66251c5bfb485c86    	       0x10bd73a61 core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h7e4107214cd3313c + 17 (function.rs:250)
17  mediasoup-66251c5bfb485c86    	       0x10ca4b2e9 std::sys::unix::thread::Thread::new::thread_start::ha4a6ad20828e5661 + 41
18  libsystem_pthread.dylib       	    0x7ff80d7d31d3 _pthread_start + 125
19  libsystem_pthread.dylib       	    0x7ff80d7cebd3 thread_start + 15

Thread 3:: SCTP iterator
0   libsystem_kernel.dylib        	    0x7ff80d7970de __psynch_cvwait + 10
1   libsystem_pthread.dylib       	    0x7ff80d7d3758 _pthread_cond_wait + 1242
2   mediasoup-66251c5bfb485c86    	       0x10c70ebe3 sctp_iterator_thread + 83
3   libsystem_pthread.dylib       	    0x7ff80d7d31d3 _pthread_start + 125
4   libsystem_pthread.dylib       	    0x7ff80d7cebd3 thread_start + 15

Thread 4:: mediasoup-worker-08c26dc2-a767-46c7-912f-f6aee2956775
0   libsystem_kernel.dylib        	    0x7ff80d7991de kevent + 10
1   mediasoup-66251c5bfb485c86    	       0x10c6ebca2 uv__io_poll + 2002
2   mediasoup-66251c5bfb485c86    	       0x10c6dfc06 uv_run + 326
3   mediasoup-66251c5bfb485c86    	       0x10c79cff5 DepLibUV::RunLoop() + 117
4   mediasoup-66251c5bfb485c86    	       0x10c7b5c4a Worker::Worker(Channel::ChannelSocket*) + 458
5   mediasoup-66251c5bfb485c86    	       0x10c7b5fbd Worker::Worker(Channel::ChannelSocket*) + 29
6   mediasoup-66251c5bfb485c86    	       0x10c7990fb mediasoup_worker_run + 1675
7   mediasoup-66251c5bfb485c86    	       0x10bea7f3f mediasoup::worker::utils::run_worker_with_channels::_$u7b$$u7b$closure$u7d$$u7d$::h09ce28fe494011ac + 1231 (utils.rs:77)
8   mediasoup-66251c5bfb485c86    	       0x10bd209ad std::sys_common::backtrace::__rust_begin_short_backtrace::h1dea32572cf1056e + 13 (backtrace.rs:135)
9   mediasoup-66251c5bfb485c86    	       0x10be628cb std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h97d78e80c9c12ecb + 43 (mod.rs:529)
10  mediasoup-66251c5bfb485c86    	       0x10bc81c4b _$LT$core..panic..unwind_safe..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h9ae2cecdcdbc05df + 43 (unwind_safe.rs:271)
11  mediasoup-66251c5bfb485c86    	       0x10bd1eaef std::panicking::try::do_call::h5ac74dcda4157391 + 63 (panicking.rs:500)
12  mediasoup-66251c5bfb485c86    	       0x10bd2092d __rust_try + 29
13  mediasoup-66251c5bfb485c86    	       0x10bd1d94c std::panicking::try::h95e60b68bdcb7520 + 92 (panicking.rs:464)
14  mediasoup-66251c5bfb485c86    	       0x10be62367 std::panic::catch_unwind::h4c18e6cce43153fe + 5 (panic.rs:142) [inlined]
15  mediasoup-66251c5bfb485c86    	       0x10be62367 std::thread::Builder::spawn_unchecked_::_$u7b$$u7b$closure$u7d$$u7d$::h3d9bc5a6e16d5de0 + 503 (mod.rs:528)
16  mediasoup-66251c5bfb485c86    	       0x10bd73a61 core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h7e4107214cd3313c + 17 (function.rs:250)
17  mediasoup-66251c5bfb485c86    	       0x10ca4b2e9 std::sys::unix::thread::Thread::new::thread_start::ha4a6ad20828e5661 + 41
18  libsystem_pthread.dylib       	    0x7ff80d7d31d3 _pthread_start + 125
19  libsystem_pthread.dylib       	    0x7ff80d7cebd3 thread_start + 15
```

ibc committed:
Add logs to file (because otherwise it's impossible to see them once the Worker is closed) and more things

Also make checker static rather than thread_local static.

Remove many debugging logs and stuff.

TODO: Rust tests do not finish.

ibc commented:
I've done many changes in https://github.com/versatica/mediasoup/pull/1353/commits/ca5f222e914cde25622de368ca927a9b3dbaf1a8:

1. Move creation of `DepUsrSCTP::Checker` from `Worker` constructor/destructor to `DepUsrSCTP::ClassInit()` and `DepUsrSCTP::ClassDestroy()`. This comes with an issue (see **Issue 2** in next comment).
2. Make the `DepUsrSCTP::Checker` singleton be `static` rather than `thread_local static`. Why? Because there must be only **one** and not one per thread (in Rust).

ibc commented:
### Issue 2

When running any test in Node:

```bash
npx jest --testPathPattern node/src/test/test-node-sctp.ts
```

```
      mediasoup:ERROR:Worker (stderr) DepLibUV::onWalk() | alive UV handle found (this shouldn't happen) [type:timer, active:0, closing:1, has_ref:1] +0ms

      492 | 			for (const line of buffer.toString('utf8').split('\n')) {
      493 | 				if (line) {
    > 494 | 					workerLogger.error(`(stderr) ${line}`);
          | 					             ^
      495 | 				}
      496 | 			}
      497 | 		});
```

This is because `DepUsrSCTP::CloseChecker()` is now called from `DepUsrSCTP::ClassDestroy()` rather than from `Worker::Close()` method. Not sure why it happens honestly, not sure what the difference is. I've added logs to file in `/tmp/ms_log.txt` and the problem is clear:

```
---- Timer created
---- Timer created
---- onCloseTimer
---- WORKER CLOSED
---- DepUsrSCTP::ClassDestroy()---- DepUsrSCTP::ClassDestroy() | calling mapAsyncHandlerSendSctpData.clear()
---- onWalk
---- onCloseTimer
```

As you can see, when `onWalk()` runs (which is defined in `DepLibUV.cpp` and it's called by `lib.cpp` when the `Worker` instance ends) the timer of the `Checker` singleton hasn't been yet freed. It's done later.

ibc committed:
make format

ibc commented:
After a meeting with Jose we have decided that, as originally planned, the timer of the usrsctp `Checker` singleton must run in a separate thread X and when the first `SctpAssociation` is created or when the last one is destroyed (in any `Worker` thread) then `DepUsrSctp` must use `uv_async_send` to tell the `Checker` singleton to start, stop or restart the timer. And same when `DepUsrSctp::HandleUsrSctpTimers()` is called if PR https://github.com/versatica/mediasoup/pull/1351 is merged.

satoren commented:
> Are you making sure that threadA is not writing to a SendSctpDataStore at the when at the same time the uv main thread is reading it?

Running a test with [ThreadSanitizer](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual) enabled may help confirm that.

ibc committed:
WIP multi thread worker usage

ibc commented:
Closing since we are about to release mediasoup with a new built-in SCTP stack.
